REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [4]:
import pandas as pd

In [7]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [8]:
bus.head()
print("The First Five Rows of the businesses DataFrame:")
for index, row in bus.head(5).iterrows():
    print(row)
    print()

Rows of the businesses DataFrame:
business_id                   10
name            TIRAMISU KITCHEN
address            033 BELDEN PL
city               San Francisco
state                         CA
postal_code                94104
latitude               37.791116
longitude            -122.403816
phone_number        +14154217044
Name: 0, dtype: object

business_id                               19
name                   NRGIZE LIFESTYLE CAFE
address         1200 VAN NESS AVE, 3RD FLOOR
city                           San Francisco
state                                     CA
postal_code                            94109
latitude                           37.786848
longitude                        -122.421547
phone_number                    +14157763262
Name: 1, dtype: object

business_id                                     24
name            OMNI S.F. HOTEL - 2ND FLOOR PANTRY
address              500 CALIFORNIA ST, 2ND  FLOOR
city                                 San Francisco
state       

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [9]:
num_rows = len(bus)
print(f"The dataset contains {num_rows} records (rows).")

The dataset contains 6315 records (rows).


#### Question 2b: How many unique business IDs are there?  

In [10]:
busunique=bus.groupby('business_id')
unique_rows = len(busunique)
print(f"The dataset contains {unique_rows} unique records (rows).")

The dataset contains 6315 unique records (rows).


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [11]:
common_names = bus['name'].value_counts().head(5)
common_names

name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [12]:
checkzip=bus.postal_code[1]
checkzip

'94109'

In [13]:
type(checkzip)

str

#### Question 3b

What are the unique values of postal_code?

In [14]:
zipcodeunique=bus.groupby('postal_code')
zipcode_rows = len(zipcodeunique)
print(f"The dataset contains {zipcode_rows} unique records (rows).")

The dataset contains 46 unique records (rows).


#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [17]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [18]:
bus_valid = bus[bus['postal_code'].isin(validZip)]
bus_valid_count = len(bus_valid)
bus_valid_count

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [20]:
missingLong=vb[pd.isnull(vb['longitude'])]
missingLong_count = len(missingLong)
missingLong_count

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [21]:
grouped = bus_valid.groupby('postal_code')['longitude'].agg(null_lon=lambda x: x.isnull().sum(), not_null_lon=lambda x: x.notnull().sum()).reset_index()
all_zip_df = pd.DataFrame({'postal_code': validZip})
zip_summary_df = pd.merge(all_zip_df, grouped, on='postal_code', how='left').fillna(0).astype({'null_lon': 'int', 'not_null_lon': 'int'})
zip_summary_df

postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [27]:
zip_summary_df['prop_missing_lon'] = zip_summary_df['null_lon'] / (zip_summary_df['null_lon'] + zip_summary_df['not_null_lon'])
top_missing_lon_zip = zip_summary_df.sort_values(by='prop_missing_lon', ascending=False).head(5)
top_missing_lon_zip[['postal_code', 'prop_missing_lon']]
##ZIP codes 94102 and 94107 appear to have more than their "fair share" of missing longitude values, with all listed businesses in these ZIP codes missing longitude data.

postal_code  prop_missing_lon
4        94107          0.547672
18       94124          0.536585
21       94132          0.485075
10       94114          0.469565
16       94122          0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [35]:
total_rows = len(ins)
unique_businesses = ins['business_id'].nunique()
total_rows, unique_businesses

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [32]:
# Counting the occurrences of each value in the 'type' column
type_counts = ins['type'].value_counts()
type_counts

type
routine      15429
complaint        1
Name: count, dtype: int64

In [34]:
# Creating a new dataframe by copying 'ins' and keeping only records with 'type' values that occur more than 10 times
types_more_than_10 = type_counts[type_counts > 10].index.tolist()
ins2 = ins[ins['type'].isin(types_more_than_10)]

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15430 rows x 4 columns]

In [36]:
# Checking the result
type_counts_result = ins2['type'].value_counts() if not ins2.empty else "No types with more than 10 occurrences"
type_counts, type_counts_result

(type
 routine      15429
 complaint        1
 Name: count, dtype: int64,
 type
 routine    15429
 Name: count, dtype: int64)

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [37]:
ins['year'] = ins['date'].astype(str).str[:4]
ins.head()

business_id  score      date     type  year
0           10     82  20160503  routine  2016
1           10     94  20140729  routine  2014
2           10     92  20140114  routine  2014
3           19     94  20160513  routine  2016
4           19     94  20141110  routine  2014

#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [42]:
# Counting the number of inspections each year
years_range = ins['year'].unique()
inspections_per_year = ins['year'].value_counts()

# Dropping records for years with less than 50 inspections
years_with_50_or_more_inspections = inspections_per_year[inspections_per_year >= 50].index
ins3 = ins[ins['year'].isin(years_with_50_or_more_inspections)]
ins3

business_id  score      date     type  year
0               10     82  20160503  routine  2016
1               10     94  20140729  routine  2014
2               10     92  20140114  routine  2014
3               19     94  20160513  routine  2016
4               19     94  20141110  routine  2014
...            ...    ...       ...      ...   ...
15425        89515     98  20161103  routine  2016
15426        89701    100  20161206  routine  2016
15427        89790     96  20161129  routine  2016
15428        89803    100  20161215  routine  2016
15429        89900     93  20161206  routine  2016

[15392 rows x 5 columns]

In [43]:
"Years with 50 or more inspections:", years_with_50_or_more_inspections.tolist()

('Years with 50 or more inspections:', ['2016', '2014', '2015'])

Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [45]:
ins_2016 = ins[ins['year'] == '2016']
ins_2016.head()

business_id  score      date     type  year
0            10     82  20160503  routine  2016
3            19     94  20160513  routine  2016
6            24     98  20161005  routine  2016
7            24     96  20160311  routine  2016
10           31     84  20160816  routine  2016

### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [50]:
merged_df = pd.merge(bus, ins_2016, on='business_id', how='left')
merged_df.head()

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4           31      NORMAN'S ICE CREAM AND FREEZES   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
4           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   

    longitude  phone_number  score        date     type  year  
0 -122.403816  +14154217044   82.0  20160503.0  routine  2016  
1 -122.421547  +14157763262   94.0  20160513.0  routine  2016  
2 -122.403135  +14156779494   98.0  20161005.0  routine  2016  
3 -122.403135  +14156779494   96.0  20160311.0  routine  2016  
4 -122.419004           NaN   84.0  20160816.0  routine  2016

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [52]:
lowest_rated = merged_df.sort_values(by='score', ascending=True).head(20)
lowest_rated_businesses = lowest_rated[['name', 'address', 'score']]
lowest_rated_businesses

name              address  score
4892      GOLDEN RIVER RESTAURANT     5827 GEARY BLVD    52.0
7325                    POKI TIME     2101 LOMBARD ST    55.0
6281                 CRAZY PEPPER   2257 SAN JOSE AVE    55.0
4873                   UNCLE CAFE       65 WAVERLY PL    55.0
3008             HING WANG BAKERY        339 JUDAH ST    56.0
5627                   GOLDEN WOK   295 B ORIZABA AVE    56.0
5383                   TOP'S CAFE  2460 SAN BRUNO AVE    57.0
6067    L & G VIETNAMESE SANDWICH         602 EDDY ST    57.0
836                  RED A BAKERY      634 CLEMENT ST    57.0
4209      SAN TUNG RESTAURANT LLC      1031 IRVING ST    57.0
549           MA'S DIM SUM & CAFE      1315 POWELL ST    57.0
4903  NEW GARDEN RESTAURANT, INC.       716 KEARNY ST    57.0
6645    KOBANI MEDITERANEAN GRILL       3242 SCOTT ST    58.0
624           JOANIE'S DINER INC.   1329 COLUMBUS AVE    59.0
6938         PAK NAM THAI CUISINE       655 LARKIN ST    59.0
3563                  JADE GARDEN      3608 BALBOA ST    59.0
4234                        ITHAI         720 POST ST    59.0
6151          VALLARTA'S TACO BAR     2022 MISSION ST    60.0
4850             HONG KONG LOUNGE     5322 GEARY BLVD    60.0
5013                 WOLFES LUNCH        1220 16TH ST    60.0

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!